### PR TITLE
Fix opam install --deps-only using the opam description of the wrong package in some cases

### LINKED ISSUE #6535: opam install --deps-only uses the wrong opam definition when building a package
Similarly to https://github.com/ocaml/opam/issues/6529, the build steps also use the wrong opam definition.
Using #6530 + #6533 + #6522 + #6520:
```
$ cat opam
opam-version: "2.0"
name: "test"
depends: ["miou"]
$ opam install .
[test] Installing new package description from upstream file:///tmp/tmp
The following actions will be performed:
=== install 2 packages
  ∗ miou 0.3.1 [required by test]
  ∗ test dev

Proceed with ∗ 1 installation? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved miou.0.3.1  (cached)
∗ installed miou.0.3.1
∗ installed test.dev
Done.

$ cat opam
opam-version: "2.0"
name: "test"
build: "false"
depends: ["miou" {= "0.3.0"}]
$ opam install --deps .
The following actions will be performed:
=== downgrade 1 package
  ↘ miou 0.3.1 to 0.3.0 [required by test]
=== recompile 1 package
  ↻ test dev (pinned)   [uses miou]

Proceed with ↘ 1 downgrade and ↻ 1 recompilation? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved miou.0.3.0  (cached)
⬇ retrieved test.dev  (file:///tmp/tmp)
⊘ removed   test.dev
⊘ removed   miou.0.3.1
∗ installed miou.0.3.0
[ERROR] The compilation of test.dev failed at "false".

#=== ERROR while compiling test.dev ===========================================#
# context     2.4.0~alpha3~dev | linux/arm64 | kit-ty-kate-platform.1 ocaml-base-compiler.5.4.0~alpha1 | pinned(file:///tmp/tmp)
# path        ~/.opam/5.4/.opam-switch/build/test.dev
# command     ~/.opam/opam-init/hooks/sandbox.sh build false
# exit-code   1
# env-file    ~/.opam/log/test-21406-cf6c92.env
# output-file ~/.opam/log/test-21406-cf6c92.out
```

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Fixes #6535

https://github.com/ocaml/opam/pull/6530 fixed the issue in the solver but created a discrepancy between what the solver gives displays and what is actually going to be executed.

To fix this we simply move the bit that re-add the non-simulated opam definitions, up before any solver/actions, and after any code specific to `--deps-only` where we actually need those information.

After careful review of the different usages of the autopin system, i think most other commands do not require fixes, at the exception of:
- the already broken `opam tree`: https://github.com/ocaml/opam/issues/6543
- possibly `opam switch create --deps-only` but i haven't managed to produce a test-case that show any issue yet. In any case i suspect this would be a very niche corner case if there is an issue there.

### CODE REVIEW

File: src/client/opamClient.ml [hunk 1]
```diff
@@ -2276,6 +2276,19 @@ let install_t t ?ask ?(ignore_conflicts=false) ?(depext_only=false)
     OpamSwitchAction.write_selections t
   );
 
+  let t =
```
Comment (rjbou) [inline src/client/opamClient.ml:2283]:
```suggestion
  (* We remove the simulated state for deps-only in order to be able to do
     actions (solve, apply) with the non simulated opam file. The simulated one
     meaningful information is already "stored" in the 'deps-of-xxx' packages
     opam files. *)
  let t =
```

Comment (kit-ty-kate) [reply src/client/opamClient.ml:2283]: applied
